### ISSUE TITLE
building deal.ii with ninja

### ISSUE THREAD

**Issue by MFraters** [dealii/dealii#9621]
I was trying to switch from building deal.ii with make to building it with ninja but I ran into a cmake error. To be clear, using the cmake + make combination works fine, so ScalaPack itselfs does not seem to be the problem. Just adding `-G 'Ninja'` causes the following error when runnign `cmake .`:
```
-- Found SCALAPACK_LIBRARY
-- Check if BLACS is embedded in ScaLAPACK library
--   SCALAPACK_LIBRARIES: /home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2020-02-27-dealii-candi/petsc-3.11.3/lib/libscalapack.a;/usr/lib/x86_64-linux-gnu/liblapack.so;/usr/lib/x86_64-linux-gnu/libblas.so;gfortran;m;quadmath;c
--   SCALAPACK_LINKER_FLAGS: 
-- Found SCALAPACK
-- Could not find a sufficient SCALAPACK installation: The required symbols pdsyevr_ and pssyevr_ were not found.
-- DEAL_II_WITH_SCALAPACK has unmet external dependencies.
CMake Error at cmake/macros/macro_configure_feature.cmake:112 (MESSAGE):
  

  Could not find the scalapack library!

  Could not find a sufficient SCALAPACK installation:

  SCALAPACK symbol check for pdsyevr_ and pssyevr_ failed! This usually means
  that your SCALAPACK installation is incomplete or the link line is broken.
  Consult

    CMakeFiles/CMakeError.log

  for further information.

  Please ensure that a suitable scalapack library is installed on your
  computer.

  If the library is not at a default location, either provide some hints for
  autodetection,

      $ SCALAPACK_DIR="..." cmake <...>
      $ cmake -DSCALAPACK_DIR="..." <...>

  or set the relevant variables by hand in ccmake.

Call Stack (most recent call first):
  cmake/macros/macro_configure_feature.cmake:269 (FEATURE_ERROR_MESSAGE)
  cmake/configure/configure_scalapack.cmake:69 (CONFIGURE_FEATURE)
  cmake/macros/macro_verbose_include.cmake:19 (INCLUDE)
  CMakeLists.txt:124 (VERBOSE_INCLUDE)
```

When I look at the `cmake + make` version (which works) the `SCALAPACK_LIBRARIES` line looks different (the part after `quadmath;m;c`, containing gcc and openmpi is missing):
```
--   SCALAPACK_LIBRARIES: /home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/petsc-3.11.3/lib/libscalapack.a;/usr/lib/x86_64-linux-gnu/liblapack.so;/usr/lib/x86_64-linux-gnu/libblas.so;gfortran;quadmath;m;c;gcc_s;gcc;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempif08.so;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempi_ignore_tkr.so;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_mpifh.so;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
```

When I look at the error message generated by cmake, it is complaning about missing mpi stuff:
```
Performing C SOURCE FILE Test DEAL_II_SCALAPACK_HAS_PDSYEVR_PSSYEVR failed with the following output:
Change Dir: /home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2020-01-24-dealii-manual/dealii/build-nj-2020-02-27/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_14248 && [1/2] Building C object CMakeFiles/cmTC_14248.dir/src.c.o
[2/2] Linking C executable cmTC_14248
FAILED: cmTC_14248 
: && /usr/bin/cc -DDEAL_II_SCALAPACK_HAS_PDSYEVR_PSSYEVR  -rdynamic CMakeFiles/cmTC_14248.dir/src.c.o  -o cmTC_14248  /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a  /usr/lib/x86_64-linux-gnu/liblapack.so  /usr/lib/x86_64-linux-gnu/libblas.so  -lgfortran  -lm  -lquadmath  -lc  /usr/lib/x86_64-linux-gnu/liblapack.so  /usr/lib/x86_64-linux-gnu/libblas.so  -lgfortran  -lm  -lquadmath  -lc && :
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(igesd2d_.o): in function `igesd2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/igesd2d_.c:74: undefined reference to `ompi_mpi_int'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/igesd2d_.c:84: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(sgesd2d_.o): in function `sgesd2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgesd2d_.c:74: undefined reference to `ompi_mpi_float'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgesd2d_.c:84: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(dgesd2d_.o): in function `dgesd2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgesd2d_.c:75: undefined reference to `ompi_mpi_double'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgesd2d_.c:85: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(igerv2d_.o): in function `igerv2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/igerv2d_.c:75: undefined reference to `ompi_mpi_int'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/igerv2d_.c:80: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(sgerv2d_.o): in function `sgerv2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgerv2d_.c:75: undefined reference to `ompi_mpi_float'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgerv2d_.c:80: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(dgerv2d_.o): in function `dgerv2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgerv2d_.c:75: undefined reference to `ompi_mpi_double'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgerv2d_.c:80: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(sgebs2d_.o): in function `sgebs2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgebs2d_.c:110: undefined reference to `ompi_mpi_float'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgebs2d_.c:117: undefined reference to `MPI_Bcast'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgebs2d_.c:118: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgebs2d_.c:189: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(dgebs2d_.o): in function `dgebs2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgebs2d_.c:110: undefined reference to `ompi_mpi_double'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgebs2d_.c:117: undefined reference to `MPI_Bcast'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgebs2d_.c:118: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgebs2d_.c:189: undefined reference to `MPI_Type_free'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(sgebr2d_.o): in function `sgebr2d_':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgebr2d_.c:120: undefined reference to `ompi_mpi_float'
...
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(sgsum2d_.oo): in function `Csgsum2d':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgsum2d_.c:156: undefined reference to `ompi_mpi_float'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgsum2d_.c:164: undefined reference to `ompi_mpi_op_sum'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgsum2d_.c:164: undefined reference to `MPI_Reduce'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgsum2d_.c:171: undefined reference to `ompi_mpi_op_sum'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/sgsum2d_.c:171: undefined reference to `MPI_Allreduce'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(dgsum2d_.oo): in function `Cdgsum2d':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgsum2d_.c:156: undefined reference to `ompi_mpi_double'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgsum2d_.c:164: undefined reference to `ompi_mpi_op_sum'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgsum2d_.c:164: undefined reference to `MPI_Reduce'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgsum2d_.c:171: undefined reference to `ompi_mpi_op_sum'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/dgsum2d_.c:171: undefined reference to `MPI_Allreduce'
/usr/bin/ld: /path/to/dealii-candi/petsc-3.11.3/lib/libscalapack.a(BI_GetMpiTrType.o): in function `BI_GetMpiTrType':
/path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/BI_GetMpiTrType.c:120: undefined reference to `MPI_Type_indexed'
/usr/bin/ld: /path/to/dealii-candi/tmp/build/petsc-lite-3.11.3/arch-linux2-c-opt/externalpackages/git.scalapack/BLACS/SRC/BI_GetMpiTrType.c:121: undefined reference to `MPI_Type_commit'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.


Source file was:

        void pdsyevr_();
        void pssyevr_();
        int main(){
          pdsyevr_();
          pssyevr_();
          return 0;
        }
```
(I removed some of the error lines with `...` and I replaced the full path with `/path/to/` for readability)

I am runninng this on ubuntu 19.10 and tested this both with `cmake version 3.16.4` and `cmake version 3.13.4`

Please let me know if you have any further questions. Thanks!

**Comment by tjhei** [comment]
Are you sure you are doing exactly the same thing? I regularly use ninja by just adding ``-G Ninja`` to the cmake command. Make sure you start from a clean build directory and double-check that you can still configure without ninja on your machine.

**Comment by MFraters** [comment]
Thanks for the quick reply :) 
Yes, to be very sure I tested it again. Here is the history of the commands:
```
 6132  mkdir build-make
 6133  cd build-make/
 6134  cmake -DTRILINOS_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/trilinos-release-12-10-1 -DSLEPC_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/slepc-3.11.2 -DSCALAPACK_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/petsc-3.11.3/lib -DPETSC_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/petsc-3.11.3 -DP4EST_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/p4est-2.2 -DOPENCASCADE_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/oce-OCE-0.18.2 -DMETIS_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/parmetis-4.0.3 -DHDF5_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/hdf5-1.10.5 -DDEAL_II_WITH_HDF5=ON -DDEAL_II_WITH_MPI=ON -DDEAL_II_WITH_P4EST=ON -DDEAL_II_WITH_PETSC=ON -DDEAL_II_WITH_SCALAPACK=ON -DDEAL_II_WITH_SLEPC=ON -DDEAL_II_WITH_THREADS=ON -DDEAL_II_WITH_TRILINOS=ON -DDEAL_II_COMPONENT_EXAMPLES=OFF ..
 6135  cd ..
 6136  mkdir build-ninja
 6137  cd build-ninja/
 6138  cmake -G 'Ninja' -DTRILINOS_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/trilinos-release-12-10-1 -DSLEPC_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/slepc-3.11.2 -DSCALAPACK_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/petsc-3.11.3/lib -DPETSC_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/petsc-3.11.3 -DP4EST_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/p4est-2.2 -DOPENCASCADE_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/oce-OCE-0.18.2 -DMETIS_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/parmetis-4.0.3 -DHDF5_DIR=/home/fraters/Documents/post-doc/2019-02-01-magali/code/dealii/2019-09-11-dealii-candi/hdf5-1.10.5 -DDEAL_II_WITH_HDF5=ON -DDEAL_II_WITH_MPI=ON -DDEAL_II_WITH_P4EST=ON -DDEAL_II_WITH_PETSC=ON -DDEAL_II_WITH_SCALAPACK=ON -DDEAL_II_WITH_SLEPC=ON -DDEAL_II_WITH_THREADS=ON -DDEAL_II_WITH_TRILINOS=ON -DDEAL_II_COMPONENT_EXAMPLES=OFF ..
```

The first generator was file, and the second showd the problem with SCALAPACK.

**Comment by bangerth** [comment]
I haven't checked that the two command lines really are exactly the same, but that would be a bizarre failure. The generator (make or ninja) has no effect on the execution of the configuration phase (or, at least, *should* not). I have no suggestion -- I've never seen anything like this before as well...

**Comment by MFraters** [comment]
Thanks for the reply. I agree that it should not make a difference in this part of the configure phase :(

I have figured out what the problem is. I noticed that the `MPI_Fortran_LIBRARIES` where empty in the ninja version, but not in the make version. I first looked whether MPI was the problem by adding `MPI_CXX_LIBRARIES` to the `DEAL_II_PACKAGE_HANDLE` in `cmake/modules/FindSCALAPACK.cmake`and although it add `/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so;/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so` to `SCALAPACK_LIBRARIES`, it did not fix the problem. 

The real problem is that ninja currently can't find Fortran to begin with: [link](https://cmake.org/cmake/help/latest/generator/Ninja.html), so that was the reason it did not fill the `MPI_Fortran_LIBRARIES` variable.

What I do not understand yet is why if SCALAPACK needs fortran, why it is not automatically disabled or asserted. Should it be, or does SCALAPACK not always require fortran?

**Comment by bangerth** [comment]
On 3/4/20 5:46 PM, Menno Fraters wrote:
> The real problem is that ninja currently can't find Fortran to begin 
> with: link <https://cmake.org/cmake/help/latest/generator/Ninja.html>, 
> so that was the reason it did not fill the |MPI_Fortran_LIBRARIES| variable.

How bizarre!

I suppose that SCALAPACK requires Fortran support libraries to be 
installed, but not necessarily a Fortran compiler. It would be nice if 
we had a way to check for the former, because that should be possible to 
set up with cmake, even if we use ninja. But then I don't know how we 
configure for SCALAPACK in our cmake scripts...

**Comment by MFraters** [comment]
I am closing this issue because the problem has been resolved starting with ninja 1.10.0 and make >= 3.7 according to the documentation. I have tested it with cmake version 3.16.3 and ninja version 1.10.0 and cmake generates fine and ninja also successfully compiles the latest master.

**Comment by tamiko** [comment]
> I have figured out what the problem is. I noticed that the MPI_Fortran_LIBRARIES where empty in the ninja version, but not in the make version.

Yes, the issue is that ninja did not support dynamic dependencies for a while (which are necessary for fortran and c++20 modules), see https://github.com/ninja-build/ninja/issues/1265
